### PR TITLE
#3037 Make @ValueMapping and @ValueMappings support combinations

### LINKED ISSUE #3037: Would it be possible to add  ElementType.ANNOTATION_TYPE as a target for @ValueMapping?
### Use case

I write a bunch of mappers for gRPC which suggests `UNSPECIFIED` as a default value for integration with languages that offer the possibility of enum members that were not predefined, and auto-generates `UNRECOGNIZED` (at least in java, it does).

I always want to map both of these with

```java
@ValueMapping(source = "UNSPECIFIED", target = MappingConstants.THROW_EXCEPTION)
@ValueMapping(source = "UNRECOGNIZED", target = MappingConstants.THROW_EXCEPTION)
```

Which, okay, isn't that big of a deal. But since I've been doing a lot of gRPC mappings, I've become accustomed to Mapping Composition. And instead of having a dozen mappings that are all ignoring various gRPC artifact fields, I extract them in their own specific annotation. 

```java
@Retention(RetentionPolicy.CLASS)
@Mapping(target = "mergeFrom", ignore = true)
@Mapping(target = "clearField", ignore = true)
@Mapping(target = "clearOneof", ignore = true)
@Mapping(target = "unknownFields", ignore = true)
@Mapping(target = "mergeUnknownFields", ignore = true)
@Mapping(target = "allFields", ignore = true)
public @interface IgnoreGrpcOnlyFields {
}
```

And then the following for a specific gRpc (proto) class:

```java
@IgnoreGrpcOnlyFields
@Mapping(target = "cityBytes", ignore = true)
@Mapping(target = "provinceBytes", ignore = true)
@Mapping(target = "zipCodeBytes", ignore = true)
// etc ...
public @interface IgnoreAddressGrpcOnlyFields {
}
```

I assume and hope that adding Mapping Composition to \@ValueMapping wouldn't represent a big technical challenge. And right now, the lack of this feature feels inconsistent. 

### Sample code

In the off case that I'm not clear enough, sample code of what I want to do:

```java
@ValueMapping(source = "UNSPECIFIED", target = MappingConstants.THROW_EXCEPTION)
@ValueMapping(source = "UNRECOGNIZED", target = MappingConstants.THROW_EXCEPTION)
public @interface IgnoreGrpcEnumFields {
}
```

```java
@IgnoreGrpcEnumFields
TargetEnum toEntity(SourceGrpcEnum source);
```
### Generated Code

Self evident, it would not change generated code, this only relates to reading the annotation metadata.

### Possible workarounds

Simply adding the individual annotations to each and every GRPC enum.

### MapStruct Version

1.5.2

### PR DESCRIPTION
**Pull request by chenzijia12300**
Fixes #3037

### CODE REVIEW

File: processor/src/test/java/org/mapstruct/ap/test/mappingcomposition/value/CustomValueAnnotation.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+import org.mapstruct.ValueMappings;
+
+/**
+ * @author orange add
+ */
+@Retention( RetentionPolicy.CLASS )
+@ValueMappings({
+    @ValueMapping(source = "EXTRA", target = "SPECIAL"),
+    @ValueMapping(source = MappingConstants.ANY_REMAINING, target = "DEFAULT")
+})
```
Comment (filiphr) [inline processor/src/test/java/org/mapstruct/ap/test/mappingcomposition/value/CustomValueAnnotation.java:22]: Let's use `ValueMapping` only here.

e.g.

```suggestion
@ValueMapping(source = "EXTRA", target = "SPECIAL"),
@ValueMapping(source = MappingConstants.ANY_REMAINING, target = "DEFAULT")
```

**Review by filiphr — CHANGES_REQUESTED** [review]
Thanks @chenzijia12300. The implementation looks good. However, I think that we need to change the location of the test a bit.

We should not touch the `CompositionTest`. Instead I would suggest creating a new `ValueCompositionTest` in a new package `org.mapstruct.ap.test.value.composition` and also put the `CustomValueAnnotation` and `ValueMappingCompositionMapper` in that package.

File: processor/src/test/java/org/mapstruct/ap/test/mappingcomposition/value/CustomValueAnnotation.java (hunk 1 above)
Comment (chenzijia12300) [reply processor/src/test/java/org/mapstruct/ap/test/mappingcomposition/value/CustomValueAnnotation.java]: Have modified

**Comment by chenzijia12300** [comment]
@filiphr Thank you for pointing it out, I have made the changes  ;D

**Review by sjaakd — APPROVED** [review]

**Comment by filiphr** [comment]
Everything looks good @chenzijia12300, I was about to merge this. However, I remembered that perhaps it will be good to add an example to the documentation as well. Can you please add a new section to the [Mapping Values](https://mapstruct.org/documentation/stable/reference/html/#mapping-enum-types) chapter?

**Comment by chenzijia12300** [comment]
@filiphr   No problem, I'll try to add an example to the documentation

**Review by sjaakd — APPROVED** [review]

**Comment by filiphr** [comment]
Thanks @chenzijia12300

**Pull request merged**